File: src/main/java/swiss/sib/sparql/playground/function/BufferCreator.java
package swiss.sib.sparql.playground.function;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.Locale;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.eclipse.rdf4j.query.algebra.evaluation.ValueExprEvaluationException;
import org.geotools.referencing.CRS;
import org.locationtech.jts.geom.Envelope;
import org.geotools.geometry.jts.JTS;
import org.opengis.geometry.Geometry;
import org.opengis.referencing.FactoryException;
import org.opengis.referencing.NoSuchAuthorityCodeException;
import org.opengis.referencing.crs.CoordinateReferenceSystem;
import org.opengis.referencing.operation.MathTransform;
import org.opengis.referencing.operation.TransformException;

public class BufferCreator {
	private static final Log logger = LogFactory.getLog(BufferCreator.class);

	private final int prec = 20;
	private final RoundingMode mode = RoundingMode.HALF_EVEN;
	private final MathContext context = new MathContext(prec, mode);

	//[m / deg]
	private final BigDecimal latUnit = BigDecimal.valueOf(111200.0).setScale(prec, mode); // consistant
	//[m / deg]
	private final BigDecimal lonUnit = BigDecimal.valueOf(87620.0).setScale(prec, mode); // California zone

	private BigDecimal x1;
	private BigDecimal y1;
	private BigDecimal x2;
	private BigDecimal y2;
	private BigDecimal distanceTotal;

	public BufferCreator(Double x1, Double y1, Double x2, Double y2, Double distance) {
		this.x1 = BigDecimal.valueOf(x1).setScale(prec, mode);
		this.y1 = BigDecimal.valueOf(y1).setScale(prec, mode);
		this.x2 = BigDecimal.valueOf(x2).setScale(prec, mode);
		this.y2 = BigDecimal.valueOf(y2).setScale(prec, mode);
		if (x2.compareTo(x1) == 0 && y2.compareTo(y1) == 0) {
			throw new ValueExprEvaluationException("Invalid LINESTRING definition: x1 equals x2 and y1 equals y2. " +
				"x1: " + x1.doubleValue() + " y1: " + y1.doubleValue() + "| x2: " + x1.doubleValue() + " y2: " + y1.doubleValue());
		}

		this.distanceTotal = BigDecimal.valueOf(distance).setScale(prec, mode);
	}

	public String create() {
		String strResult = "POINT EMPTY";

		try {
			if (x2.compareTo(x1) == 0) {
				strResult = equalLongitudesSpecialCase();

			} else if (y2.compareTo(y1) == 0) {
				strResult = equalLatitudesSpecialCase();

			} else {
				strResult = basicCase();
			}

		} catch (Exception e) {
			logger.error(e.getMessage(), e);
		}

		return strResult;
	}

	private String formatPolygonStr(BigDecimal pX1, BigDecimal pY1, BigDecimal pX2, BigDecimal pY2, BigDecimal pX3,
			BigDecimal pY3, BigDecimal pX4, BigDecimal pY4) {

		return String.format(Locale.US, "POLYGON ((%.15f %.15f, %.15f %.15f, %.15f %.15f, %.15f %.15f, %.15f %.15f))",
				pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4, pX1, pY1);
	}

	// CASE x1 = x2
	private String equalLongitudesSpecialCase() {
		// correcting the order
		if (y2.doubleValue() < y1.doubleValue()) {
			BigDecimal temp = y2;
			y2 = y1;
			y1 = temp;
		}

		// TODO: UMT conversion
		// latitude change [deg]: latDeg = distanceTotal / latUnit
		BigDecimal latDeg = distanceTotal.divide(latUnit, context); // latUnit will never be zero
		// longitude change [deg]: lonDeg = distanceTotal / lonUnit
		BigDecimal lonDeg = distanceTotal.divide(lonUnit, context); // lonUnit will never be zero

		// POLYGONE POINTS

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		BigDecimal pX1 = x1.subtract(lonDeg, context);
		BigDecimal pY1 = y1.subtract(latDeg, context);

		// pX2 = x1 + lonDeg
		// pY2 = y1 - latDeg
		BigDecimal pX2 = x1.add(lonDeg, context);
		BigDecimal pY2 = y1.subtract(latDeg, context);

		// pX3 = x2 + lonDeg
		// pY3 = y2 + latDeg
		BigDecimal pX3 = x2.add(lonDeg, context);
		BigDecimal pY3 = y2.add(latDeg, context);

		// pX4 = x2 - lonDeg
		// pY4 = y2 + latDeg
		BigDecimal pX4 = x2.subtract(lonDeg, context);
		BigDecimal pY4 = y2.add(latDeg, context);

		return formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
	}

	// CASE y1 = y2
	private String equalLatitudesSpecialCase() {
		// correcting the order
		if (x2.doubleValue() < x1.doubleValue()) {
			BigDecimal temp = x2;
			x2 = x1;
			x1 = temp;
		}

		// TODO: UMT conversion
		// latitude change [deg]: latDeg = distanceTotal / latUnit
		BigDecimal latDeg = distanceTotal.divide(latUnit, context); // latUnit will never be zero
		// longitude change [deg]: lonDeg = distanceTotal / lonUnit
		BigDecimal lonDeg = distanceTotal.divide(lonUnit, context); // lonUnit will never be zero

		// POLYGONE POINTS

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		BigDecimal pX1 = x1.subtract(lonDeg, context);
		BigDecimal pY1 = y1.subtract(latDeg, context);

		// pX2 = x1 - lonDeg
		// pY2 = y1 + latDeg
		BigDecimal pX2 = x1.subtract(lonDeg, context);
		BigDecimal pY2 = y1.add(latDeg, context);

		// pX3 = x2 + lonDeg
		// pY3 = y2 + latDeg
		BigDecimal pX3 = x2.add(lonDeg, context);
		BigDecimal pY3 = y2.add(latDeg, context);

		// pX4 = x2 + lonDeg
		// pY4 = y2 - latDeg
		BigDecimal pX4 = x2.add(lonDeg, context);
		BigDecimal pY4 = y2.subtract(latDeg, context);

		return formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
	}

	// CASE x1 != x2 && y1 != y2
	private String basicCase() throws Exception {
		// slope of acline segment: slope = (y2 - y1) / (x2 - x1)
		BigDecimal slope = y2.subtract(y1, context).divide(x2.subtract(x1, context), context); // x2 and x1 will not be equal
		Boolean isSharp = slope.compareTo(BigDecimal.valueOf(0).setScale(prec, mode)) > 0;

		// puts coordinates in right order
		correctCoordOrder(isSharp);
		// extends line for distanceTotal length at each end
		extendLine(slope);

		return createPolygonStr(slope);
	}

	private void correctCoordOrder(Boolean isSharp) {
		Boolean correctionNeeded = false;

		if (x1.compareTo(x2) > 0) {
			correctionNeeded = true;
		}

		if (isSharp && y1.compareTo(y2) > 0) {
			correctionNeeded = true;
		}

		if (!isSharp && y1.compareTo(y2) < 0) {
			correctionNeeded = true;
		}

		if (correctionNeeded) {
			BigDecimal temp = x1;
			x1 = x2;
			x2 = temp;

			temp = y1;
			y1 = y2;
			y2 = temp;
		}
	}

	// extends the line for distanceTotal at each end
	private void extendLine(BigDecimal slope) throws NoSuchAuthorityCodeException, FactoryException, TransformException {
		// angle of line to x-axis [rad]: angle = arcus tangent(slope)
		BigDecimal angle = BigDecimal.valueOf(Math.atan(slope.doubleValue())).setScale(prec, mode);
		if (angle.doubleValue() == 0 || ((Double) angle.doubleValue()).isNaN()) {
			throw new ValueExprEvaluationException("perpendicular angle: " + angle.doubleValue());
		}

		// X and Y components of total distance [m]:
		// distanceY = distanceTotal * sin(angle)
		BigDecimal angleSin = BigDecimal.valueOf(Math.sin(angle.doubleValue())).setScale(prec, mode);
		BigDecimal distanceY = distanceTotal.multiply(angleSin, context);
		// distanceX = distanceY / pSlope
		BigDecimal distanceX = distanceY.divide(slope, context); // slope will not be zero if latitudes are not equal

		// TODO: UMT conversion
		CoordinateReferenceSystem sourceCRS = CRS.decode("EPSG:4326");
		CoordinateReferenceSystem targetCRS = CRS.decode("EPSG:25832");
		MathTransform transform = CRS.findMathTransform(sourceCRS, targetCRS, false);
		Geometry targetGeometry = (Geometry) JTS.transform(new Envelope(), transform);
		
		// longitude change [deg]
		BigDecimal lonDeg = distanceX.divide(lonUnit, context); // lonUnit will never be zero
		// latitude change [deg]
		BigDecimal latDeg = distanceY.divide(latUnit, context); // latUnit will never be zero
		
		// x1 = x1 - lonDeg
		// y1 = y1 - latDeg
		x1 = x1.subtract(lonDeg, context);
		y1 = y1.subtract(latDeg, context);

		// x2 = x2 + lonDeg
		// y2 = y2 + latDeg
		x2 = x2.add(lonDeg, context);
		y2 = y2.add(latDeg, context);
	}

	private String createPolygonStr(BigDecimal slope) throws Exception {
		// slope of perpendicular line: pSlope = - 1 / Slope
		BigDecimal pSlope = BigDecimal.valueOf(-1).setScale(prec, mode).divide(slope, context); // slope will not be zero if latitudes are not equal

		// angle of perpendicular line to x-axis [rad]: pAngle = arcus tangent(pSlope)
		BigDecimal pAngle = BigDecimal.valueOf(Math.atan(pSlope.doubleValue())).setScale(prec, mode);
		if (pAngle.doubleValue() == 0 || ((Double) pAngle.doubleValue()).isNaN()) {
			throw new ValueExprEvaluationException("perpendicular angle: " + pAngle.doubleValue());
		}

		// X and Y components of total distance [m]:
		// distanceY = distanceTotal * sin(pAngle)
		BigDecimal angleSin = BigDecimal.valueOf(Math.sin(pAngle.doubleValue())).setScale(prec, mode);
		BigDecimal distanceY = distanceTotal.multiply(angleSin, context);
		// distanceX = distanceY / pSlope
		BigDecimal distanceX = distanceY.divide(pSlope, context); // slope will not be zero if latitudes are not equal

		// TODO: UMT conversion
		// longitude change [deg]
		BigDecimal lonDeg = distanceX.divide(lonUnit, context); // lonUnit will never be zero
		// latitude change [deg]
		BigDecimal latDeg = distanceY.divide(latUnit, context); // latUnit will never be zero
		
		// POLYGONE POINTS
		BigDecimal pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4;

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		pX1 = x1.subtract(lonDeg, context);
		pY1 = y1.subtract(latDeg, context);

		// pX1 = x1 + lonDeg
		// pY1 = y1 + latDeg
		pX2 = x1.add(lonDeg, context);
		pY2 = y1.add(latDeg, context);

		// pX1 = x2 + lonDeg
		// pY1 = y2 + latDeg
		pX3 = x2.add(lonDeg, context);
		pY3 = y2.add(latDeg, context);

		// pX1 = x2 - lonDeg
		// pY1 = y2 - latDeg
		pX4 = x2.subtract(lonDeg, context);
		pY4 = y2.subtract(latDeg, context);

		return formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
	}
}

File: src/main/java/swiss/sib/sparql/playground/function/javascript/buffer-creator.mjs
const BigNumber = require('/bignumber.js');

export class BufferCreator {
    constructor(x1, y1, x2, y2, distanceTotal) {
		this.latUnit = new BigNumber(111200.0);
		this.lonUnit = new BigNumber(78630.0);

		this.x1 = new BigNumber(x1);
		this.y1 = new BigNumber(y1);
		this.x2 = new BigNumber(x2);
		this.y2 = new BigNumber(y2);
		this.distanceTotal = new BigNumber(distanceTotal);
    }

    evaluate() {
		var strResult = "";
		if (this.x2.isEqualTo(this.x1)) {
			strResult = this.equalLatitudesSpecialCase();

		} else if (this.y2.isEqualTo(this.y1)) {
			strResult = this.equalLongitudesSpecialCase();

		} else {
			strResult = this.basicCase();
		}

		return strResult;
    }

    // CASE x1 = x2
    equalLatitudesSpecialCase() {
		if (this.y2.isLessThan(this.y1)) {
			let temp = this.y2;
			this.y2 = this.y1;
			this.y1 = temp;
		}

		// TODO: UMT conversion
		// latitude change [deg]: latDeg = distanceTotal / latUnit
		let latDeg = this.distanceTotal.dividedBy(this.latUnit); // latUnit will never be zero
		// longitude change [deg]: lonDeg = distanceTotal / lonUnit
		let lonDeg = this.distanceTotal.dividedBy(this.lonUnit); // lonUnit will never be zero

		// POLYGONE POINTS

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		let pX1 = this.x1.minus(lonDeg);
		let pY1 = this.y1.minus(latDeg);

		// pX2 = x1 + lonDeg
		// pY2 = y1 - latDeg
		let pX2 = this.x1.plus(lonDeg);
		let pY2 = this.y1.minus(latDeg);

		// pX3 = x2 + lonDeg
		// pY3 = y2 + latDeg
		let pX3 = this.x2.plus(lonDeg);
		let pY3 = this.y2.plus(latDeg);

		// pX4 = x2 - lonDeg
		// pY4 = y2 + latDeg
		let pX4 = this.x2.minus(lonDeg);
		let pY4 = this.y2.plus(latDeg);

		return this.formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
    }

    // CASE y1 = y2
    equalLongitudesSpecialCase() {
		if (this.x2.isLessThan(this.x1)) {
			let temp = this.x2;
			this.x2 = this.x1;
			this.x1 = temp;
		}

		// TODO: UMT conversion
		// latitude change [deg]: latDeg = distanceTotal / latUnit
		let latDeg = this.distanceTotal.dividedBy(this.latUnit); // latUnit will never be zero
		// longitude change [deg]: lonDeg = distanceTotal / lonUnit
		let lonDeg = this.distanceTotal.dividedBy(this.lonUnit); // lonUnit will never be zero

		// POLYGONE POINTS

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		let pX1 = this.x1.minus(lonDeg);
		let pY1 = this.y1.minus(latDeg);

		// pX2 = x1 - lonDeg
		// pY2 = y1 + latDeg
		let pX2 = this.x1.minus(lonDeg);
		let pY2 = this.y1.plus(latDeg);

		// pX3 = x2 + lonDeg
		// pY3 = y2 + latDeg
		let pX3 = this.x2.plus(lonDeg);
		let pY3 = this.y2.plus(latDeg);

		// pX4 = x2 + lonDeg
		// pY4 = y2 - latDeg
		let pX4 = this.x2.plus(lonDeg);
		let pY4 = this.y2.minus(latDeg);

		return this.formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
    }

    // CASE x1 != x2 && y1 != y2
    basicCase() {
      	// slope of acline segment: slope = (y2 - y1) / (x2 - x1)
		let slope = this.y2.minus(this.y1).dividedBy(this.x2.minus(this.x1)); // x2 and x1 will not be equal
		let isSharp = slope.isPositive();

      	// puts coordinates in right order
		this.correctCoordOrder(isSharp);
		// extends line for distanceTotal length at each end
		this.extendLine(slope);

		return this.createPolygonStr(slope);
    }

    correctCoordOrder(isSharp) {
		let correctionNeeded = false;

		if (this.x1.isGreaterThan(this.x2)) {
			correctionNeeded = true;
		}

		if (isSharp && this.y1.isGreaterThan(this.y2)) {
			correctionNeeded = true;
		}

		if (!isSharp && this.y1.isLessThan(this.y2)) {
			correctionNeeded = true;
		}

		if (correctionNeeded) {
			let temp = this.x1;
			this.x1 = this.x2;
			this.x2 = temp;

			temp = this.y1;
			this.y1 = this.y2;
			this.y2 = temp;
		}
    }

    // extends the line for distanceTotal at each end
    extendLine(slope) {
		// angle of line to x-axis [rad]: pAngle = arcus tangent(pSlope)
		let angle = new BigNumber(Math.atan(slope));
		if (angle.isZero() || angle.isNaN()) {
			throw new Error("perpendicular angle: " + angle.doubleValue());
		}

		// X and Y components of total distance [m]:
		// distanceY = distanceTotal * sin(angle)
		let angleSin = new BigNumber(Math.sin(angle));
		let distanceY = this.distanceTotal.multipliedBy(angleSin);
		// distanceX = distanceY / pSlope
		let distanceX = distanceY.dividedBy(slope); // slope will not be zero if latitudes are not equal

		// TODO: UMT conversion
		// latitude change [deg]
		let latDeg = distanceX.dividedBy(this.latUnit); // latUnit will never be zero
		// longitude change [deg]
		let lonDeg = distanceY.dividedBy(this.lonUnit); // lonUnit will never be zero

		// x1 = x1 - lonDeg
		// y1 = y1 - latDeg
		this.x1 = this.x1.minus(lonDeg);
		this.y1 = this.y1.minus(latDeg);

		// x2 = x2 + lonDeg
		// y2 = y2 + latDeg
		this.x2 = this.x2.plus(lonDeg);
		this.y2 = this.y2.plus(latDeg);
    }

    createPolygonStr(slope) {
      	// slope of perpendicular line: pSlope = - 1 / Slope
		let pSlope = new BigNumber(-1).dividedBy(slope); // slope will not be zero if longitudes are not equal

		// angle of perpendicular line to x-axis [rad]: pAngle = arcus tangent(pSlope)
		let pAngle = new BigNumber(Math.atan(pSlope));
		if (pAngle.isZero() || pAngle.isNaN()) {
			throw new Error("perpendicular angle: " + pAngle.doubleValue());
		}

		// X and Y components of total distance [m]:
		// distanceY = distanceTotal * sin(angle)
		let angleSin = new BigNumber(Math.sin(pAngle));
		let distanceY = this.distanceTotal.multipliedBy(angleSin);
		// distanceX = distanceY / pSlope
		let distanceX = distanceY.dividedBy(pSlope); // slope will not be zero if latitudes are not equal

		// TODO: UMT conversion
		// latitude change [deg]
		let latDeg = distanceX.dividedBy(this.latUnit); // latUnit will never be zero
		// longitude change [deg]
		let lonDeg = distanceY.dividedBy(this.lonUnit); // lonUnit will never be zero

		// pX1 = x1 - lonDeg
		// pY1 = y1 - latDeg
		let pX1 = this.x1.minus(lonDeg);
		let pY1 = this.y1.minus(latDeg);

		// pX2 = x1 + lonDeg
		// pY2 = y1 + latDeg
		let pX2 = this.x1.plus(lonDeg);
		let pY2 = this.y1.plus(latDeg);

		// pX3 = x2 + lonDeg
		// pY3 = y2 + latDeg
		let pX3 = this.x2.plus(lonDeg);
		let pY3 = this.y2.plus(latDeg);

		// pX4 = x2 - lonDeg
		// pY4 = y2 - latDeg
		let	pX4 = this.x2.minus(lonDeg);
		let pY4 = this.y2.minus(latDeg);

		return this.formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4);
    }

    formatPolygonStr(pX1, pY1, pX2, pY2, pX3, pY3, pX4, pY4) {
		let str = "";
		str = str.concat("POLYGON ((")
				.concat(pX1).concat(" ").concat(pY1).concat(", ")
				.concat(pX2).concat(" ").concat(pY2).concat(", ")
				.concat(pX3).concat(" ").concat(pY3).concat(", ")
				.concat(pX4).concat(" ").concat(pY4).concat(", ")
				.concat(pX1).concat(" ").concat(pY1).concat("))");
		return str;
    }
}
